Treat some exceptions as unhandled when a debugger is attached

goderbauer opened the pull request:
Fixes https://github.com/flutter/flutter/issues/17007

This does elevate the experience when using an interactive debugger.

The following bugs in the Dart SDK prevent us from using the `vm:notify-debugger-on-exception` in all possible instances. I've added TODOs in the code for those cases.
* https://github.com/dart-lang/sdk/issues/45673 (async methods)
* https://github.com/dart-lang/sdk/issues/45710 (named closures)
* https://github.com/dart-lang/sdk/issues/45684 (debugger not breaking on correct line)

/cc @devoncarew @DanTup

goderbauer committed:
++

goderbauer committed:
++

goderbauer committed:
++

goderbauer committed:
++

goderbauer committed:
++

DanTup commented:
> However, I am a little concerned about the maintainability of this. How will anybody know (or even if they do, remember) to add the pragma to newly created methods that swallow exceptions? Any ideas?

I don't know if a good solution to this. Being able to detect it automatically would require knowing when `catch` blocks are catching exceptions likely in the end users code, and that's probably not easy to tell.

Do all of these functions call `FlutterError.reportError`? If so, there may be some value in adding a note to the dartdoc for that indicating that methods that call it in catch blocks should probably have the pragma. Whether that could be a lint, I'm not sure (I don't think there's a way to have project-specific lints right now - at least without an analysis server plugin).

goderbauer commented:
I've already included a note in the dart docs:

https://github.com/flutter/flutter/pull/78649/files#diff-6c2151db0538ec4ea5b79ae6d89e1430fc6bb3278c499710f456d32e49fc423aR1103

devoncarew commented:
> However, I am a little concerned about the maintainability of this. How will anybody know (or even if they do, remember) to add the pragma to newly created methods that swallow exceptions? Any ideas?

I don't have a good answer to that. I was assuming that there would be a ~central place - in frame building code or other - where errors were caught. We could either annotate that method or perform some refactoring so that more of the uncaught framework errors flowed through that method.

This annotation needed to be added to more places than I was assuming it would be. I think we still want to proceed forward - it's a good benefit for the user. If people come up with good ideas to make using the annotation w/ the framework more maintainable, I'm happy to help out w/ any implementation costs.

cc @jacob314 @csells for FYI that this work - better exceptions w/ Flutter code - is very close to landing

devoncarew commented:
This look great!

FYI, there are several debugger (and VM service protocol) integration tests in packages/flutter_tools (see for example packages/flutter_tools/test/integration.shard/debugger_stepping_test.dart). These tests could be a good model for how to write a test which verifies that an uncaught exception does in fact break.

@jonahwilliams might also have a good idea about how to test this.

jonahwilliams commented:
I would definitely look at some of the intergration.shard tests - those will attach to an app running in the flutter_tester device, which you could then connect to via VM Service/exercise the exception throwing a bit

goderbauer commented:
Quick update: I am in the process of adding tests to it using the suggested mechanism. It's just a slow process...

jonahwilliams commented:
@goderbauer if you'd like to sync up later tomorrow/this week I'm happy to help

goderbauer commented:
Also, the testing uncovered that the pragma doesn't appear to work on `async` methods: https://github.com/flutter/flutter/issues/17007#issuecomment-818318520

goderbauer committed:
first batch of tests (11/19)

goderbauer committed:
cleanup

goderbauer committed:
update comment

goderbauer committed:
more tests (16/20)

goderbauer committed:
more tests (20/20)

goderbauer commented:
Added tests that should cover all occurrences. There are 4 instances in which the pragma didn't work:

1. `async` methods (like `_DefaultBinaryMessenger.handlePlatformMessage`), issue: https://github.com/dart-lang/sdk/issues/45673
2. When the try/catch is in an anonymous closure (as seen in `_LayoutBuilderElement._layout`), asked for clarification in https://github.com/flutter/flutter/issues/17007#issuecomment-818952818
3. In the following instances the pragma caused the debugger to break, but the break happened on an unrelated (random?) line. I haven't figured out yet what's different about these methods that is causing this:
    * `BuildOwner.finalizeTree`
    * `SchedulerBinding.handleEventLoopCallback`
    
I've removed the pragma from all locations where it didn't work as expected (and replaced it with a TODO). I think we can fix those locations up in a follow-up PR.

With that, PTAL @devoncarew @jonahwilliams @DanTup.

goderbauer committed:
whitespace

goderbauer commented:
I think the behavior I am describing under point 3 in the enumeration above is unrelated to the pragma. It appears to be a bug in Dart, I have filed https://github.com/dart-lang/sdk/issues/45684, which contains steps to reproduce the issue outside of Flutter.

jonahwilliams commented:
Wow those are quite the tests :) 

We may need to shard a little bit harder for those, but i'd let CI run and see what the elapsed time is first.

jonahwilliams commented on packages/flutter_tools/test/integration.shard/break_on_framework_exceptions_test.dart:
```diff
@@ -0,0 +1,737 @@
+    await flutter.test(withDebugger: true, pauseOnExceptions: true);
+    await flutter.waitForPause();
+
+    expect((await flutter.getSourceLocation()).line, equals(project.lineContaining(project.test, "throw 'AnimationController listener';")));
```
nit: equals doesn't really do much here besides make the lines longer. Also, I would split these lines up a bit to make this easier to read. perhaps something like:

```dart
final IDonTKnowWhatType sourceLocation = await flutter.getSourceLocation();
expect(sourceLocation.line, project.lineContaining(project.test, "throw 'AnimationController listener';"));

````

goderbauer committed:
update TODOs

goderbauer commented:
> We may need to shard a little bit harder for those, but i'd let CI run and see what the elapsed time is first.

Locally, `flutter test test/integration.shard/break_on_framework_exceptions_test.dart  ` finishes in a little over 4 minutes.

goderbauer committed:
jonah review

goderbauer commented:
@jonahwilliams Looks like it is pushing the Windows shard over the 30 min timeout? Should we create a forth tool_integration_tests shard?

jonahwilliams commented:
definitely, I think we setup the shards so that is fairly easy to add. FYI @jmagman

jmagman commented:
> definitely, I think we setup the shards so that is fairly easy to add. FYI @jmagman

https://github.com/flutter/infra/pull/412

goderbauer committed:
update TODO for LayoutBuilder

goderbauer commented:
> flutter/infra#412

Thanks, @jmagman!

goderbauer commented:
This is ready for review. Landing it will require an infra update (https://github.com/flutter/infra/pull/412) first.

In a follow-up PR we can address use cases that are currently blocked due to the following bugs in the Dart SDK (I've added TODO in the appropriate places in this PR):
* https://github.com/dart-lang/sdk/issues/45673 (async methods)
* https://github.com/dart-lang/sdk/issues/45710 (named closures)
* https://github.com/dart-lang/sdk/issues/45684 (debugger not breaking on correct line)

jonahwilliams reviewed (approved):
LGTM
